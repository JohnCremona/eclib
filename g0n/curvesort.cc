// FILE CURVESORT.CC:  isogeny class id codes etc
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2007 John Cremona
// 
// This file is part of the mwrank/g0n package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////

#define USE_NEW_CODE_LETTERS

int BA[] = {1,0};
int ACB[] = {0,2,1};
int BAC[] = {1,0,2};
int BCA[] = {1,2,0};
int CAB[] = {2,0,1};
int CBA[] = {2,1,0};
int ABDC[] = {0,1,3,2};
int ACBD[] = {0,2,1,3};
int ACDB[] = {0,2,3,1};
int ADBC[] = {0,3,1,2};
int ADCB[] = {0,3,2,1};
int BACD[] = {1,0,2,3};
int BADC[] = {1,0,3,2};
int BCAD[] = {1,2,0,3};
int BCDA[] = {1,2,3,0};
int BDAC[] = {1,3,0,2};
int CABD[] = {2,0,1,3};
int CBAD[] = {2,1,0,3};
int CDAB[] = {2,3,0,1};
int CDBA[] = {2,3,1,0};
int DABC[] = {3,0,1,2};
int DACB[] = {3,0,2,1};
int DBAC[] = {3,1,0,2};
int DCAB[] = {3,2,0,1};
int AEBDC[] = {0,4,1,3,2};
int AECBD[] = {0,4,2,1,3};
int AEDCB[] = {0,4,3,2,1};
int BCADE[] = {1,2,0,3,4};
int BCAED[] = {1,2,0,4,3};
int BEACD[] = {1,4,0,2,3};
int BECAD[] = {1,4,2,0,3};
int CEADB[] = {2,4,0,3,1};
int CABDE[] = {2,0,1,3,4};
int CEDAB[] = {2,4,3,0,1};
int CEDBA[] = {2,4,3,1,0};
int DAECB[] = {3,0,4,2,1};
int DCAEB[] = {3,2,0,4,1};
int DBCAE[] = {3,1,2,0,4};
int EABDC[] = {4,0,1,3,2};
int EACBD[] = {4,0,2,1,3};
int EBCAD[] = {4,1,2,0,3};
int EDACB[] = {4,3,0,2,1};
int ADECFB[] = {0,3,4,2,5,1};
int AECDBF[] = {0,4,2,3,1,5};
int BADECF[] = {1,0,3,4,2,5};
int BDAECF[] = {1,3,0,4,2,5};
int BFADCE[] = {1,5,0,3,2,4};
int CABEDF[] = {2,0,1,4,3,5};
int CAFBDE[] = {2,0,5,1,3,4};
int CEABDF[] = {2,4,0,1,3,5};
int DABFEC[] = {3,0,1,5,4,2};
int DCABEF[] = {3,2,0,1,4,5};
int EABCDF[] = {4,0,1,2,3,5};
int EBACDF[] = {4,1,0,2,3,5};
int ECABDF[] = {4,2,0,1,3,5};
int AEDBGFC[] = {0,4,3,1,6,5,2};
int AFGDEBC[] = {0,5,6,3,4,1,2};
int ECFBDGA[] = {4,2,5,1,3,6,0};
int EFCAGDB[] = {4,5,2,0,6,3,1};
int EFGCABD[] = {4,5,6,2,0,1,3};
int FGACBED[] = {5,6,0,2,1,4,3};
int FGBDACE[] = {5,6,1,3,0,2,4};
int FGDBAEC[] = {5,6,3,1,0,4,2};
int FGDBEAC[] = {5,6,3,1,4,0,2};
int FGDEABC[] = {5,6,3,4,0,1,2};
int AGBDEFHC[] = {0,6,1,3,4,5,7,2};
int AFHCGDEB[] = {0,5,7,2,6,3,4,1};
int BECADFGH[] = {1,4,2,0,3,5,6,7};
int EGBAFHCD[] = {4,6,1,0,5,7,2,3};
int GBEAHDFC[] = {6,1,4,0,7,3,5,2};

int booknumber0(int level, int form)  // utes number starting from 0
{
  if(level<56) return form;
  if(level>450) return form;
switch (level) {
case 56: return BA[form]; break;
case 77: return ACB[form]; break;
case 84: return BA[form]; break;
case 99: return ACBD[form]; break;
case 102: return ACB[form]; break;
case 106: return DACB[form]; break;
case 110: return CBA[form]; break;
case 114: return CAB[form]; break;
case 116: return CAB[form]; break;
case 118: return ACDB[form]; break;
case 120: return BA[form]; break;
case 121: return CABD[form]; break;
case 123: return BA[form]; break;
case 124: return BA[form]; break;
case 126: return BA[form]; break;
case 128: return ADCB[form]; break;
case 130: return ACB[form]; break;
case 132: return BA[form]; break;
case 136: return BA[form]; break;
case 140: return BA[form]; break;
case 141: return EBCAD[form]; break;
case 142: return EABDC[form]; break;
case 144: return BA[form]; break;
case 147: return ACB[form]; break;
case 150: return CAB[form]; break;
case 153: return ADCB[form]; break;
case 154: return ACB[form]; break;
case 155: return CBA[form]; break;
case 158: return EACBD[form]; break;
case 162: return ACBD[form]; break;
case 168: return BA[form]; break;
case 170: return DAECB[form]; break;
case 171: return ADBC[form]; break;
case 174: return CEDBA[form]; break;
case 175: return CAB[form]; break;
case 178: return BA[form]; break;
case 182: return CEADB[form]; break;
case 184: return DABC[form]; break;
case 185: return BCA[form]; break;
case 186: return ACB[form]; break;
case 187: return BA[form]; break;
case 189: return ADBC[form]; break;
case 190: return BAC[form]; break;
case 192: return ACBD[form]; break;
case 195: return CDBA[form]; break;
case 196: return BA[form]; break;
case 198: return CEDAB[form]; break;
case 200: return BECAD[form]; break;
case 201: return BCA[form]; break;
case 203: return BCA[form]; break;
case 205: return CAB[form]; break;
case 208: return DABC[form]; break;
case 210: return DBCAE[form]; break;
case 212: return BA[form]; break;
case 214: return DBAC[form]; break;
case 219: return BCA[form]; break;
case 221: return BA[form]; break;
case 222: return EDACB[form]; break;
case 234: return BCADE[form]; break;
case 235: return CAB[form]; break;
case 236: return BA[form]; break;
case 238: return DAECB[form]; break;
case 240: return BCAD[form]; break;
case 242: return BA[form]; break;
case 245: return CAB[form]; break;
case 246: return EFCAGDB[form]; break;
case 249: return BA[form]; break;
case 252: return BA[form]; break;
case 254: return DCAB[form]; break;
case 256: return BACD[form]; break;
case 262: return BA[form]; break;
case 264: return DABC[form]; break;
case 267: return BA[form]; break;
case 270: return BCDA[form]; break;
case 272: return ADBC[form]; break;
case 274: return CBA[form]; break;
case 278: return BA[form]; break;
case 285: return CAB[form]; break;
case 286: return BDAECF[form]; break;
case 288: return AECBD[form]; break;
case 291: return DBAC[form]; break;
case 294: return EFGCABD[form]; break;
case 297: return DACB[form]; break;
case 298: return BA[form]; break;
case 300: return ACDB[form]; break;
case 302: return BAC[form]; break;
case 304: return ECABDF[form]; break;
case 306: return CBAD[form]; break;
case 312: return CAFBDE[form]; break;
case 315: return BA[form]; break;
case 318: return DCAEB[form]; break;
case 320: return EABCDF[form]; break;
case 322: return BADC[form]; break;
case 324: return ABDC[form]; break;
case 325: return EABDC[form]; break;
case 326: return ACB[form]; break;
case 330: return AEDCB[form]; break;
case 333: return DABC[form]; break;
case 336: return CABEDF[form]; break;
case 338: return ADECFB[form]; break;
case 339: return BAC[form]; break;
case 342: return FGDEABC[form]; break;
case 345: return AECDBF[form]; break;
case 348: return BDAC[form]; break;
case 350: return BFADCE[form]; break;
case 352: return CEABDF[form]; break;
case 354: return DCABEF[form]; break;
case 360: return BCAED[form]; break;
case 364: return BA[form]; break;
case 366: return FGBDACE[form]; break;
case 368: return AEDBGFC[form]; break;
case 369: return BA[form]; break;
case 370: return ACBD[form]; break;
case 372: return ACDB[form]; break;
case 378: return GBEAHDFC[form]; break;
case 380: return BA[form]; break;
case 381: return BA[form]; break;
case 384: return BECADFGH[form]; break;
case 387: return CABDE[form]; break;
case 390: return AFGDEBC[form]; break;
case 392: return EBACDF[form]; break;
case 396: return BAC[form]; break;
case 400: return AFHCGDEB[form]; break;
case 402: return ABDC[form]; break;
case 404: return BA[form]; break;
case 405: return DABFEC[form]; break;
case 406: return ACDB[form]; break;
case 408: return ADCB[form]; break;
case 410: return BDAC[form]; break;
case 414: return CDAB[form]; break;
case 418: return BAC[form]; break;
case 423: return ECFBDGA[form]; break;
case 425: return ADBC[form]; break;
case 426: return CAB[form]; break;
case 427: return BCA[form]; break;
case 432: return EGBAFHCD[form]; break;
case 434: return AEBDC[form]; break;
case 435: return BACD[form]; break;
case 437: return BA[form]; break;
case 438: return FGACBED[form]; break;
case 440: return ADBC[form]; break;
case 441: return BADECF[form]; break;
case 442: return BEACD[form]; break;
case 446: return ADCB[form]; break;
case 448: return AGBDEFHC[form]; break;
case 450: return FGDBEAC[form]; break;

 default: return form;
}
  return form;  //default case for levels not yet sorted manually
}


const char alphabet[] = "abcdefghijklmnopqrstuvwxyz";

int booknumber(int level, int form)  // permutes number starting from 1
{
  return 1+booknumber0(level,form-1);
}

// Old codes were:  A,B,...,Z,AA,BB,...,ZZ,AAA,BBB,... etc

// Function to convert old code to integer (from 0)

int old_codeletter_to_int(char* code)  // i counts from 0!
{
  int i = code[0]-'A';
  int n=1; while(code[n]!='\0') n++;
  return 26*(n-1)+i;
}

// Function to convert integer (from 0) to old code

void old_codeletter(int i, char* code, int width=0)
{
  int n=width;    // pads string to this width with blanks
  code[n]='\0';
  while (n) code[--n]=' ';

  int nc = i%26;
  char c = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"[nc];
  n = 1 + (i-nc)/26;
  if(width==0) code[n]='\0';
  while (n) code[--n]=c;
  /*
  int j = old_codeletter_to_int(code);
  if(i==j) return;
  cout<<i<<" -> "<<code<<" -> "<<j<<endl;
  */
}

// old-new codes were:  a,b,...,z,aa,ab,...,az,ba,bb,... etc

// Function to convert old-new code to integer (from 0)

int old_new_codeletter_to_int(char* code)  // i counts from 0!
{
  int b = code[0]-'a';
  if(code[1]=='\0')  return b;
  int a = code[1]-'a';
  return 26*(b+1)+a;
}

// Function to convert integer (from 0) to old-new code

void old_new_codeletter(int i, char* code)  // i counts from 0!
{
  int b = i%26;
  int a = (i-b)/26;
  if (a==0) {code[0]=alphabet[b]; code[1]='\0';}
  else {code[0]=alphabet[a-1]; code[1]=alphabet[b]; code[2]='\0';}
  /*
  int j = codeletter_to_int(code);
  if(i==j) return;
  cout<<i<<" -> "<<code<<" -> "<<j<<endl;
  */
}

// new-new codes (from 01.08.05) are:  a,b,...,z,ba,bb,...,bz,ca,cb,... etc.,  i.e. straight base 26 with digits a=0, b=1, ..., z=25

// Function to convert new code to integer (from 0)

int codeletter_to_int(char* code)  // i counts from 0!
{
  int b = code[0]-'a';
  if(code[1]=='\0')  return b;
  int a = code[1]-'a';
  return 26*b+a;
}

// Function to convert integer (from 0) to new code

void new_codeletter(int i, char* code)  // i counts from 0!
{
  int b = i%26;
  int a = (i-b)/26;
  if (a==0) {code[0]=alphabet[b]; code[1]='\0';}
  else {code[0]=alphabet[a]; code[1]=alphabet[b]; code[2]='\0';}
  /*
  int j = codeletter_to_int(code);
  if(i==j) return;
  cout<<i<<" -> "<<code<<" -> "<<j<<endl;
  */
}

#ifdef USE_NEW_CODE_LETTERS
inline void codeletter(int i, char* code) {return new_codeletter(i,code);}
#else
inline void codeletter(int i, char* code) {return old_codeletter(i,code);}
#endif

