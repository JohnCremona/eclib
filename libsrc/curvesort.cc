// FILE CURVESORT.CC:  isogeny class id codes etc
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

#include <sstream>
#include <algorithm> // for reverse()
#include "eclib/curvesort.h"

int curvesort_BA[] = {1,0};
int curvesort_ACB[] = {0,2,1};
int curvesort_BAC[] = {1,0,2};
int curvesort_BCA[] = {1,2,0};
int curvesort_CAB[] = {2,0,1};
int curvesort_CBA[] = {2,1,0};
int curvesort_ABDC[] = {0,1,3,2};
int curvesort_ACBD[] = {0,2,1,3};
int curvesort_ACDB[] = {0,2,3,1};
int curvesort_ADBC[] = {0,3,1,2};
int curvesort_ADCB[] = {0,3,2,1};
int curvesort_BACD[] = {1,0,2,3};
int curvesort_BADC[] = {1,0,3,2};
int curvesort_BCAD[] = {1,2,0,3};
int curvesort_BCDA[] = {1,2,3,0};
int curvesort_BDAC[] = {1,3,0,2};
int curvesort_CABD[] = {2,0,1,3};
int curvesort_CBAD[] = {2,1,0,3};
int curvesort_CDAB[] = {2,3,0,1};
int curvesort_CDBA[] = {2,3,1,0};
int curvesort_DABC[] = {3,0,1,2};
int curvesort_DACB[] = {3,0,2,1};
int curvesort_DBAC[] = {3,1,0,2};
int curvesort_DCAB[] = {3,2,0,1};
int curvesort_AEBDC[] = {0,4,1,3,2};
int curvesort_AECBD[] = {0,4,2,1,3};
int curvesort_AEDCB[] = {0,4,3,2,1};
int curvesort_BCADE[] = {1,2,0,3,4};
int curvesort_BCAED[] = {1,2,0,4,3};
int curvesort_BEACD[] = {1,4,0,2,3};
int curvesort_BECAD[] = {1,4,2,0,3};
int curvesort_CEADB[] = {2,4,0,3,1};
int curvesort_CABDE[] = {2,0,1,3,4};
int curvesort_CEDAB[] = {2,4,3,0,1};
int curvesort_CEDBA[] = {2,4,3,1,0};
int curvesort_DAECB[] = {3,0,4,2,1};
int curvesort_DCAEB[] = {3,2,0,4,1};
int curvesort_DBCAE[] = {3,1,2,0,4};
int curvesort_EABDC[] = {4,0,1,3,2};
int curvesort_EACBD[] = {4,0,2,1,3};
int curvesort_EBCAD[] = {4,1,2,0,3};
int curvesort_EDACB[] = {4,3,0,2,1};
int curvesort_ADECFB[] = {0,3,4,2,5,1};
int curvesort_AECDBF[] = {0,4,2,3,1,5};
int curvesort_BADECF[] = {1,0,3,4,2,5};
int curvesort_BDAECF[] = {1,3,0,4,2,5};
int curvesort_BFADCE[] = {1,5,0,3,2,4};
int curvesort_CABEDF[] = {2,0,1,4,3,5};
int curvesort_CAFBDE[] = {2,0,5,1,3,4};
int curvesort_CEABDF[] = {2,4,0,1,3,5};
int curvesort_DABFEC[] = {3,0,1,5,4,2};
int curvesort_DCABEF[] = {3,2,0,1,4,5};
int curvesort_EABCDF[] = {4,0,1,2,3,5};
int curvesort_EBACDF[] = {4,1,0,2,3,5};
int curvesort_ECABDF[] = {4,2,0,1,3,5};
int curvesort_AEDBGFC[] = {0,4,3,1,6,5,2};
int curvesort_AFGDEBC[] = {0,5,6,3,4,1,2};
int curvesort_ECFBDGA[] = {4,2,5,1,3,6,0};
int curvesort_EFCAGDB[] = {4,5,2,0,6,3,1};
int curvesort_EFGCABD[] = {4,5,6,2,0,1,3};
int curvesort_FGACBED[] = {5,6,0,2,1,4,3};
int curvesort_FGBDACE[] = {5,6,1,3,0,2,4};
int curvesort_FGDBAEC[] = {5,6,3,1,0,4,2};
int curvesort_FGDBEAC[] = {5,6,3,1,4,0,2};
int curvesort_FGDEABC[] = {5,6,3,4,0,1,2};
int curvesort_AGBDEFHC[] = {0,6,1,3,4,5,7,2};
int curvesort_AFHCGDEB[] = {0,5,7,2,6,3,4,1};
int curvesort_BECADFGH[] = {1,4,2,0,3,5,6,7};
int curvesort_EGBAFHCD[] = {4,6,1,0,5,7,2,3};
int curvesort_GBEAHDFC[] = {6,1,4,0,7,3,5,2};

int booknumber0(int level, int form)  // permutes numbers starting from 0
{
  if(level<56) return form;
  if(level>450) return form;
switch (level) {
case 56: return curvesort_BA[form]; break;
case 77: return curvesort_ACB[form]; break;
case 84: return curvesort_BA[form]; break;
case 99: return curvesort_ACBD[form]; break;
case 102: return curvesort_ACB[form]; break;
case 106: return curvesort_DACB[form]; break;
case 110: return curvesort_CBA[form]; break;
case 114: return curvesort_CAB[form]; break;
case 116: return curvesort_CAB[form]; break;
case 118: return curvesort_ACDB[form]; break;
case 120: return curvesort_BA[form]; break;
case 121: return curvesort_CABD[form]; break;
case 123: return curvesort_BA[form]; break;
case 124: return curvesort_BA[form]; break;
case 126: return curvesort_BA[form]; break;
case 128: return curvesort_ADCB[form]; break;
case 130: return curvesort_ACB[form]; break;
case 132: return curvesort_BA[form]; break;
case 136: return curvesort_BA[form]; break;
case 140: return curvesort_BA[form]; break;
case 141: return curvesort_EBCAD[form]; break;
case 142: return curvesort_EABDC[form]; break;
case 144: return curvesort_BA[form]; break;
case 147: return curvesort_ACB[form]; break;
case 150: return curvesort_CAB[form]; break;
case 153: return curvesort_ADCB[form]; break;
case 154: return curvesort_ACB[form]; break;
case 155: return curvesort_CBA[form]; break;
case 158: return curvesort_EACBD[form]; break;
case 162: return curvesort_ACBD[form]; break;
case 168: return curvesort_BA[form]; break;
case 170: return curvesort_DAECB[form]; break;
case 171: return curvesort_ADBC[form]; break;
case 174: return curvesort_CEDBA[form]; break;
case 175: return curvesort_CAB[form]; break;
case 178: return curvesort_BA[form]; break;
case 182: return curvesort_CEADB[form]; break;
case 184: return curvesort_DABC[form]; break;
case 185: return curvesort_BCA[form]; break;
case 186: return curvesort_ACB[form]; break;
case 187: return curvesort_BA[form]; break;
case 189: return curvesort_ADBC[form]; break;
case 190: return curvesort_BAC[form]; break;
case 192: return curvesort_ACBD[form]; break;
case 195: return curvesort_CDBA[form]; break;
case 196: return curvesort_BA[form]; break;
case 198: return curvesort_CEDAB[form]; break;
case 200: return curvesort_BECAD[form]; break;
case 201: return curvesort_BCA[form]; break;
case 203: return curvesort_BCA[form]; break;
case 205: return curvesort_CAB[form]; break;
case 208: return curvesort_DABC[form]; break;
case 210: return curvesort_DBCAE[form]; break;
case 212: return curvesort_BA[form]; break;
case 214: return curvesort_DBAC[form]; break;
case 219: return curvesort_BCA[form]; break;
case 221: return curvesort_BA[form]; break;
case 222: return curvesort_EDACB[form]; break;
case 234: return curvesort_BCADE[form]; break;
case 235: return curvesort_CAB[form]; break;
case 236: return curvesort_BA[form]; break;
case 238: return curvesort_DAECB[form]; break;
case 240: return curvesort_BCAD[form]; break;
case 242: return curvesort_BA[form]; break;
case 245: return curvesort_CAB[form]; break;
case 246: return curvesort_EFCAGDB[form]; break;
case 249: return curvesort_BA[form]; break;
case 252: return curvesort_BA[form]; break;
case 254: return curvesort_DCAB[form]; break;
case 256: return curvesort_BACD[form]; break;
case 262: return curvesort_BA[form]; break;
case 264: return curvesort_DABC[form]; break;
case 267: return curvesort_BA[form]; break;
case 270: return curvesort_BCDA[form]; break;
case 272: return curvesort_ADBC[form]; break;
case 274: return curvesort_CBA[form]; break;
case 278: return curvesort_BA[form]; break;
case 285: return curvesort_CAB[form]; break;
case 286: return curvesort_BDAECF[form]; break;
case 288: return curvesort_AECBD[form]; break;
case 291: return curvesort_DBAC[form]; break;
case 294: return curvesort_EFGCABD[form]; break;
case 297: return curvesort_DACB[form]; break;
case 298: return curvesort_BA[form]; break;
case 300: return curvesort_ACDB[form]; break;
case 302: return curvesort_BAC[form]; break;
case 304: return curvesort_ECABDF[form]; break;
case 306: return curvesort_CBAD[form]; break;
case 312: return curvesort_CAFBDE[form]; break;
case 315: return curvesort_BA[form]; break;
case 318: return curvesort_DCAEB[form]; break;
case 320: return curvesort_EABCDF[form]; break;
case 322: return curvesort_BADC[form]; break;
case 324: return curvesort_ABDC[form]; break;
case 325: return curvesort_EABDC[form]; break;
case 326: return curvesort_ACB[form]; break;
case 330: return curvesort_AEDCB[form]; break;
case 333: return curvesort_DABC[form]; break;
case 336: return curvesort_CABEDF[form]; break;
case 338: return curvesort_ADECFB[form]; break;
case 339: return curvesort_BAC[form]; break;
case 342: return curvesort_FGDEABC[form]; break;
case 345: return curvesort_AECDBF[form]; break;
case 348: return curvesort_BDAC[form]; break;
case 350: return curvesort_BFADCE[form]; break;
case 352: return curvesort_CEABDF[form]; break;
case 354: return curvesort_DCABEF[form]; break;
case 360: return curvesort_BCAED[form]; break;
case 364: return curvesort_BA[form]; break;
case 366: return curvesort_FGBDACE[form]; break;
case 368: return curvesort_AEDBGFC[form]; break;
case 369: return curvesort_BA[form]; break;
case 370: return curvesort_ACBD[form]; break;
case 372: return curvesort_ACDB[form]; break;
case 378: return curvesort_GBEAHDFC[form]; break;
case 380: return curvesort_BA[form]; break;
case 381: return curvesort_BA[form]; break;
case 384: return curvesort_BECADFGH[form]; break;
case 387: return curvesort_CABDE[form]; break;
case 390: return curvesort_AFGDEBC[form]; break;
case 392: return curvesort_EBACDF[form]; break;
case 396: return curvesort_BAC[form]; break;
case 400: return curvesort_AFHCGDEB[form]; break;
case 402: return curvesort_ABDC[form]; break;
case 404: return curvesort_BA[form]; break;
case 405: return curvesort_DABFEC[form]; break;
case 406: return curvesort_ACDB[form]; break;
case 408: return curvesort_ADCB[form]; break;
case 410: return curvesort_BDAC[form]; break;
case 414: return curvesort_CDAB[form]; break;
case 418: return curvesort_BAC[form]; break;
case 423: return curvesort_ECFBDGA[form]; break;
case 425: return curvesort_ADBC[form]; break;
case 426: return curvesort_CAB[form]; break;
case 427: return curvesort_BCA[form]; break;
case 432: return curvesort_EGBAFHCD[form]; break;
case 434: return curvesort_AEBDC[form]; break;
case 435: return curvesort_BACD[form]; break;
case 437: return curvesort_BA[form]; break;
case 438: return curvesort_FGACBED[form]; break;
case 440: return curvesort_ADBC[form]; break;
case 441: return curvesort_BADECF[form]; break;
case 442: return curvesort_BEACD[form]; break;
case 446: return curvesort_ADCB[form]; break;
case 448: return curvesort_AGBDEFHC[form]; break;
case 450: return curvesort_FGDBEAC[form]; break;

 default: return form;
}
  return form;  //default case for levels not yet sorted manually
}

int booknumber(int level, int form)  // permutes numbers starting from 1
{
  return 1+booknumber0(level,form-1);
}

// new-new codes (from 01.08.05) are:
// a,b,...,z,ba,bb,...,bz,ca,cb,... etc., i.e. straight base 26 with
// digits a=0, b=1, ..., z=25

// Function to convert new code to integer (from 0) for any length of code.
int codeletter_to_int(string code)  // i counts from 0!
{
  int n=0;
  std::for_each(code.begin(), code.end(), [&n] ( const char& c ) { n*=26; n+=(c-'a');});
  return n;
}

// Function to convert integer (from 0) to new code

string new_codeletter(int i)  // i counts from 0!
{
  if (i==0) return string("a"); // special case -- otherwise leading
                                // a's are omitted
  stringstream code;
  int n = i;
  while (n)
  {
    std::div_t x = div(n,26);
    code << alphabet[x.rem];
    n = x.quot;
  }
  string res = code.str();
  reverse(res.begin(),res.end());
  return res;
}

