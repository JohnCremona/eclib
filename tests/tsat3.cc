// tsat3.cc -- test for saturate.h/cc reading from gens files directly
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
//

#include "interface.h"
#include "matrix.h"
#include "curve.h"
#include "points.h"
#include "cperiods.h"
#include "polys.h"
#include "curvemod.h"
#include "pointsmod.h"
#include "saturate.h"
#include "elog.h"
#include "sieve_search.h"
#include "mwprocs.h"
#include "curvesort.cc"

#define USE_EGR
#define SAT_BND 100000 // saturation bound:  use -1 for global default

#ifdef USE_EGR
const int use_egr=1;
#else
const int use_egr=0;
#endif

#define INPUT_CLASS_IS_LETTER // we only use letters now!

//void codeletter(int i, char* code, int width=0);
// Utility function for parsing input of lists of integers such as [], [2], [2,2] 
vector<int> input_list(istream & is);

int main()
{
  //  set_precision(string("Enter number of decimal places").c_str());
  set_precision(100);
  initprimes(string("PRIMES").c_str(),0);
  int verbose = 0;
  //  cout<<"verbose (0/1)? ";             cin >>verbose;
  int j, npts;

  long N, ncurve;
  char code[20];
  Curve E;

  while(!(feof(stdin))) {

  // Input the curve's ID and the curve:

  cin >> N; if(N==0) break;
#ifdef INPUT_CLASS_IS_LETTER
  cin >> code;
#else
  long nclass;
  cin >> nclass; 
  codeletter((nclass-1),code);
#endif
  cin >> ncurve;
  cin >> E;
  Curvedata C(E);
  cout<<endl;
  cout << N<<code<<ncurve<<" = "<< E << endl;

  // Input the rank:

  Point P(C);
  cin >> npts;
  cout<<"rank =  "<<npts<<endl;

  // Input the torsion structure:
  vector<int> torsion_group = input_list(cin);
  int trank = torsion_group.size();
  cout<<"torsion group structure =  "<<torsion_group<<" (torsion rank "<<trank<<")"<<endl;

  // Input the points of infinite order (if any):

  vector<Point> points; points.reserve(npts);
  j=0; 
  while(j<npts)
    { 
      cin >> P;
      if ( !P.isvalid() ) 
	{
	  cout<<"point "<<P<<" not on curve.\n\n"; 
	  abort();
	}
      points.push_back(P); 
      j++;
    }
  cout<<"Input non-torsion points: "<<points<<endl;

  // Input the points of finite order (if any):

  vector<Point> tpoints; points.reserve(trank);
  j=0; 
  while(j<trank)
    { 
      cin >> P;
      if ( !P.isvalid() ) 
	{
	  cout<<"point "<<P<<" not on curve.\n\n"; 
	  abort();
	}
      tpoints.push_back(P); 
      j++;
    }
  cout<<"Input torsion points: "<<tpoints<<endl;

  bigint index;
  vector<long> unsatprimes;
  int success = saturate_points(C, points, index, unsatprimes, SAT_BND, use_egr, (verbose));

  if(success)
    {
      cout<<"Saturation complete --";
      if(index==1) 
	cout<<"input points were saturated"<<endl;
      else
	{
	  cout<<"input points had index "<<index
	      <<" in their saturation."<<endl;
	  cout<<"Basis for saturation:\t"<<points<<endl;
	}
    }
  else
    cout<<"Saturation failed at "<<unsatprimes<<endl;

  for (int i=0; i<npts; i++)
    { 
      Point P = points[i];
      cout << "Generator "<<(i+1)<<" is "<<P<<"; ";
      cout << "height "<<height(P);
      if(!P.isvalid()) cout<<" --warning: NOT on curve!";
      cout<<endl;
    }
  cout<<endl;
  cout << "Regulator = "<<regulator(points)<<endl<<endl;

  // Finally output a line similar to the input line:

#ifdef INPUT_CLASS_IS_LETTER
  cout<<N<<"\t"<<code<<"\t"<<ncurve<<"\t"<<E<<"\t"<<npts;
#else
  cout<<N<<"\t"<<nclass<<"\t"<<ncurve<<"\t"<<E<<"\t"<<npts;
#endif
  for(j=0; j<npts; j++) cout<<"\t"<<points[j];
  cout<<endl;
  cout<<"=============================================================="<<endl;
  }
}

// Utility function for parsing input of lists of integers such as [], [2], [2,2] 
vector<int> input_list(istream & is)
{
  char c; int a; vector<int> ai;
  is>>c;  // swallow first [
  is>>ws>>c;
  if (c==']') return ai;
  is.unget();
  while (c!=']')
    {
      is >> a >> c; // c is a comma or ]
      ai.push_back(a);
   }
  return ai; 
}




#if(0)
void codeletter(int i, char* code, int width)
{
  int n=width;    // pads string to this width with blanks
  code[n]='\0';
  while (n) code[--n]=' ';

  int nc = i%26;
  char c = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"[nc];
  n = 1 + (i-nc)/26;
  if(width==0) code[n]='\0';
  while (n) code[--n]=c;
}
#endif

//end of file tsat3.cc

