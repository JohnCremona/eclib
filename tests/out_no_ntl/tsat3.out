
24129181 = [0,0,1,-68799,-1969790]
rank =  2
torsion group structure =  [ ] (torsion rank 0)
Input non-torsion points: [ [-1870:9257:8] [-197:1984:1] ]
Input torsion points: [ ]
Saturation complete --input points were saturated
Generator 1 is [-1870:9257:8]; height 2.00839
Generator 2 is [-197:1984:1]; height 0.182405

Regulator = 0.288919

24129	18	1	[0,0,1,-68799,-1969790]	2	[-1870:9257:8]	[-197:1984:1]
==============================================================

133507b2 = [1,-1,0,-898,-7905]
rank =  2
torsion group structure =  [ 2 2 ] (torsion rank 2)
Input non-torsion points: [ [-178:299:8] [-23748:-59351:1728] ]
Input torsion points: [ [34:-17:1] [-10:5:1] ]
Saturation complete --input points were saturated
Generator 1 is [-178:299:8]; height 4.69293
Generator 2 is [-23748:-59351:1728]; height 6.99486

Regulator = 32.7933

133507	b	2	[1,-1,0,-898,-7905]	2	[-178:299:8]	[-23748:-59351:1728]
==============================================================
